onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h

```diff
@@ -134,6 +132,52 @@ Status CheckInputs(const Tensor* query,
                            "Input 'past_key' and 'past_value' shall be both present or both absent");
   }
 
+  if (key != nullptr) {
+    const auto& key_dims = key->Shape().GetDims();
+    if (key_dims.size() != 3) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'key' is expected to have 3 dimensions, got ",
+                             key_dims.size());
+    }
+    if (query_dims[0] != key_dims[0]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Input 'query' and 'key' shall have same dim 0 (batch size)");
+    }
+
+    if (num_heads % kv_num_heads != 0) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "num_heads must be a multiple of kv_num_heads. Got num_heads % kv_num_heads == ",
+                             num_heads % kv_num_heads);
+    }
+  } else {
```

Comment:
You do not need check missing key. That will be validated by operator schema (key is not optional).